### PR TITLE
Components: Add HOF to ignore IME keydowns

### PR DESCRIPTION
**Pull request by mirka**
## What?

Extracts the IME-handing logic we added in #46827 etc. into a higher-order function for easier reuse.

### Affected components

- Autocomplete
- ComboboxControl
- FormTokenField
- Modal

## Why?

For code DRYness and clarity, as well as in preparation for fixing #58188 which will likely require another one of these.

This kind of HOF doesn't seem necessary outside of the components package yet, but we might want to consider moving it to `@wordpress/compose` if we start to see more use cases. Currently the only usage is [in the Command Palette](https://github.com/WordPress/gutenberg/blob/888c78e16212afd3ff08148512387bbdb9132e76/packages/commands/src/components/command-menu.js#L256-L262) — although I'm not exactly sure if it's still needed, since there's already a handler in the Modal. ~In any case the Escape handling in there is broken right now (#59080).~

## Testing Instructions for Keyboard

Smoke test the affected components to make sure that they're ignoring IME events as before.

For example, open a `Modal` in the Storybook and use a CJK IME to type in the text field inside the dialog. <kbd>Esc</kbd> keypresses during a composition should not close the modal.

### CODE REVIEW

**Review by t-hamano — COMMENTED** [review]
Thanks for the PR! I think it makes sense to create a HOF because if we have an issue with the IME in the future, we'll probably write the same code.

By the way, I was not able to reproduce the problem reported in #59080 in my environment. It might be a good idea to first identify the environment in which the issue reported in #59080 occurs.

**Review by tyxla — COMMENTED** [review]
This is a neat abstraction, thanks 👍 

I prefer leaving a full review to someone who uses an input method editor more frequently, but this looks good code-wise!

One thing I was curious about is whether we can add some unit tests for the new HOF?

**Comment by mirka** [comment]
> It might be a good idea to first identify the environment in which the issue reported in [#59080](https://github.com/WordPress/gutenberg/issues/59080) occurs.

@t-hamano I've determined the cause of #59080 to be a certain Chrome extension. I don't know what changed in the Escape handling in Command Palette from WP 6.4 to 6.5, and any other Escape handling is still working, such as in the Modal component. Anyway, I've disabled the extension in my wp-admins for now.

> One thing I was curious about is whether we can add some unit tests for the new HOF?

@tyxla Not really, at least in an easy or reliable way. There is no official support for IME composition in [Puppeteer](https://github.com/puppeteer/puppeteer/issues/4981) nor [Playwright](https://github.com/microsoft/playwright/issues/5777) yet. It seems like you can now [simulate composition](https://github.com/quilljs/quill/commit/0ea789f95fc4956287b3995f9495aafa367d4190) with some new Chrome APIs and whatnot, but not sure how meaningful it will be for this specific use case. Maybe useful to test basic IME compatibility in more complex applications like the editor writing flow.

At this stage, I think we'd have a bigger impact on QA if we focused on developer awareness, for example adding IME info to the keyboard testing guides for a11y, or adding some kind of linter rule for `onKeyDown` stuff. I'll do a keydown handler audit in the components package and see what we can do.

**Comment by t-hamano** [comment]
> I've determined the cause of #59080 to be a certain Chrome extension. I don't know what changed in the Escape handling in Command Palette from WP 6.4 to 6.5, and any other Escape handling is still working, such as in the Modal component. Anyway, I've disabled the extension in my wp-admins for now.

Great to hear that! If so, is it possible to delete this part as well?

https://github.com/WordPress/gutenberg/blob/96cb434382c2a70d991eab85ceb542b76c30a94a/packages/commands/src/components/command-menu.js#L255-L266

**Comment by mirka** [comment]
> is it possible to delete this part as well?

I'd like that to be a separate PR, since it warrants some extra investigation/testing, and it'll be easier to revert if something goes wrong 👍

File: packages/components/src/modal/index.tsx [hunk 1]
```diff
@@ -371,6 +361,7 @@ function UnforwardedModal(
  * ```jsx
  * import { Button, Modal } from '@wordpress/components';
  * import { useState } from '@wordpress/element';
+  import { withIgnoreIMEEvents } from '../utils/with-ignore-ime-events';
```
Comment (t-hamano) [inline packages/components/src/modal/index.tsx]: This is unnecessary, right?

**Review by t-hamano — APPROVED** [review]
> I'd like that to be a separate PR, since it warrants some extra investigation/testing, and it'll be easier to revert if something goes wrong 👍

I see, then I think it's okay to ship this PR. I have already confirmed that all affected components continue to work correctly 👍

I only left feedback in one place.

File: packages/components/src/modal/index.tsx (hunk 1 above)
Comment (mirka) [reply packages/components/src/modal/index.tsx]: Oops, VS Code sometimes gets confused and auto-imports stuff in the JSDoc code snippet 😅 Fixed!

**Pull request merged**